skglm/utils/validation.py

```diff
@@ -23,3 +28,68 @@ def check_group_compatible(obj):
                 f"'{obj_name}' is not block-separable. "
                 f"Missing '{attr}' attribute."
             )
+
+
+def check_obj_solver_attr(obj, solver, required_attr, support_sparse=False):
+    """Check whether datafit or penalty is compatible with solver.
+
+    Parameters
+    ----------
+    obj : Instance of Datafit or Penalty
+        The instance Datafit (or Penalty) to check.
+
+    solver : Instance of Solver
+        The instance of Solver to check.
+
+    required_attr : List or tuple of strings
+        The attributes that ``obj`` must have.
+
+    support_sparse : bool, default False
+        If ``True`` adds a ``SPARSE_SUFFIX`` to check compatibility with sparse data.
+
+    Raises
+    ------
+        AttributeError
+            if any of the attribute in ``required_attr`` is missing
+            from ``obj`` attributes.
+    """
+    missing_attrs = []
+    suffix = SPARSE_SUFFIX if support_sparse else ""
+
+    # if `attr` is a list check that at least one of them
+    # is within `obj` attributes
+    for attr in required_attr:
+        attributes = attr if not isinstance(attr, str) else (attr,)
+
+        for a in attributes:
+            if hasattr(obj, f"{a}{suffix}"):
+                break
+        else:
+            missing_attrs.append(_join_attrs_with_or(attributes, suffix))
+
+    if len(missing_attrs):
+        required_attr = [_join_attrs_with_or(attrs, suffix) for attrs in required_attr]
+
+        # get name obj and solver
+        name_matcher = re.compile(r"\.(\w+)'>")
+
+        obj_name = name_matcher.search(str(obj.__class__)).group(1)
+        solver_name = name_matcher.search(str(solver.__class__)).group(1)
+
+        raise AttributeError(
+            f"{obj_name} is not compatible with {solver_name}"
```

Comment:
I find this hard to read, and I don't get why we prefix the sentence with a `.` is not `support_sparse`.
Two separate ifs are less fancy but may be more readable.